File: src/CandidatosCore.Core/Historial/RegistroCambio.cs
namespace CandidatosCore.Core.Historial;

using CandidatosCore.Core.Dominio;
using CandidatosCore.Core.Fases;

public sealed class RegistroCambio
{
    public Candidato Candidato { get; }
    public FaseBase EstadoAnterior { get; }
    public FaseBase EstadoNuevo { get; }
    public string Actor { get; }
    public DateTime Fecha { get; }
    public TipoCambio Tipo { get; }

    public string Comentario { get; }
    public bool EsNotaInterna { get; }

    public RegistroCambio(Candidato candidato, FaseBase estadoAnterior, FaseBase estadoNuevo, string actor, TipoCambio tipo, string comentario = "", bool esNotaInterna = false)
    {
        Candidato = candidato;
        EstadoAnterior = estadoAnterior;
        EstadoNuevo = estadoNuevo;
        Actor = actor;
        Tipo = tipo;
        Fecha = DateTime.UtcNow;
        Comentario = comentario;
        EsNotaInterna = esNotaInterna;
    }
}

File: src-ts/CandidatosCore.Core/Historial/RegistroCambio.ts
import type { Candidato } from "../Dominio/Candidato";
import type { FaseBase } from "../Fases";
import { TipoCambio } from "./TipoCambio";

export class RegistroCambio {
    public readonly candidato: Candidato;
    public readonly estadoAnterior: FaseBase;
    public readonly estadoNuevo: FaseBase;
    public readonly actor: string;
    public readonly fecha: Date;
    public readonly tipo: TipoCambio;
    public readonly comentario?: string;

    constructor(
        candidato: Candidato,
        estadoAnterior: FaseBase,
        estadoNuevo: FaseBase,
        actor: string,
        tipo: TipoCambio,
        comentario?: string
    ) {
        this.candidato = candidato;
        this.estadoAnterior = estadoAnterior;
        this.estadoNuevo = estadoNuevo;
        this.actor = actor;
        this.tipo = tipo;
        this.fecha = new Date();
        this.comentario = comentario;
    }
}
